File: java-code/src/main/java/refactoring/conditional_logic/decompose_conditional/after/ChargeCalculator.java
package refactoring.conditional_logic.decompose_conditional.after;

import java.time.LocalDate;

public class ChargeCalculator {

  public double calculateCharge(LocalDate date, double quantity, Plan plan) {

    //STEP 4: Reformat the if statement with a ternary operator
    return summer(date, plan) ? summerCharge(quantity, plan) : regularCharge(quantity, plan);
  }

  //STEP 3: Extract else logic to a method
  private double regularCharge(double quantity, Plan plan) {
    return quantity * plan.getRegularRate() + plan.getRegularServiceCharge();
  }

  //STEP 2: Extract if logic to a method
  private double summerCharge(double quantity, Plan plan) {
    return quantity * plan.getSummerRate();
  }

  //STEP 1: Extract if validation logic to a method
  private boolean summer(LocalDate aDate, Plan plan) {
    return !aDate.isBefore(plan.getSummerStart()) && !aDate.isAfter(plan.getSummerEnd());
  }
}

File: go-code/conditional_logic/decompose_conditional/after/charge_calculator.go
package before

import (
	"time"
)

type ChargeCalculator struct{}

func (cc *ChargeCalculator) calculateCharge(aDate time.Time, quantity float64, plan Plan) float64 {

	//STEP 4: Refactor if/else
	if cc.summer(aDate, plan) {
		return cc.summerCharge(quantity, plan)
	} else {
		return cc.regularCharge(quantity, plan)
	}
}

// STEP 3: Extract else logic to a method
func (cc *ChargeCalculator) regularCharge(quantity float64, plan Plan) float64 {
	return quantity*plan.RegularRate + plan.RegularServiceCharge
}

// STEP 2: Extract if logic to a method
func (cc *ChargeCalculator) summerCharge(quantity float64, plan Plan) float64 {
	return quantity * plan.SummerRate
}

// STEP 1: Extract if validation logic to a method
func (cc *ChargeCalculator) summer(aDate time.Time, plan Plan) bool {
	return !aDate.Before(plan.SummerStart) && !aDate.After(plan.SummerEnd)
}
